src/dawn/Optimizer/PassMultiStageSplitter.cpp

```diff
@@ -27,101 +27,169 @@
+
+        if(conflict.CounterLoopOrderConflict) {
+
+          // The loop order of the lower part is what we recoreded in the last steps
+          splitterIndices.emplace_front(
+              MultiStage::SplitIndex{stageIndex, stmtIndex, curLoopOrder});
+
+          if(context->getOptions().ReportPassMultiStageSplit)
+            std::cout << "\nPASS: " << name << ": " << stencilInstantiation->getName() << ": split:"
+                      << doMethod.getStatementAccessesPairs()[stmtIndex]
+                             ->getStatement()
+                             ->ASTStmt->getSourceLocation()
+                             .Line
+                      << " looporder:" << curLoopOrder << "\n";
+
+          if(context->getOptions().DumpSplitGraphs)
+            graph.toDot(format("stmt_vd_ms%i_%02i.dot", multiStageIndex, numSplit));
+
+          // Clear the graph ...
+          graph.clear();
+          curLoopOrder = LoopOrderKind::LK_Parallel;
+
+          // ... and process the current statement again
+          graph.insertStatementAccessesPair(stmtAccessesPair);
 
-      // If not proven otherwise, we assume a parralel loop order
-      auto curLoopOrder = LoopOrderKind::LK_Parallel;
+          numSplit++;
 
-      // Build the Dependency graph (bottom --> top i.e iterate the stages backwards)
-      int stageIndex = multiStage.getStages().size() - 1;
-      for(auto stageIt = multiStage.getStages().rbegin(); stageIt != multiStage.getStages().rend();
-          ++stageIt, --stageIndex) {
-        Stage& stage = (**stageIt);
-        DoMethod& doMethod = stage.getSingleDoMethod();
+        } else if(conflict.LoopOrderConflict)
+          // We have a conflict in the loop order, the multi-stage cannot be executed in
+          // parallel and we use the loop order specified by the user
+          curLoopOrder = userSpecifiedLoopOrder;
+      }
+    }
 
-        // Iterate statements backwards
-        for(int stmtIndex = doMethod.getStatementAccessesPairs().size() - 1; stmtIndex >= 0;
-            --stmtIndex) {
-          auto& stmtAccessesPair = doMethod.getStatementAccessesPairs()[stmtIndex];
-          graph.insertStatementAccessesPair(stmtAccessesPair);
+    if(context->getOptions().DumpSplitGraphs)
+      graph.toDot(format("stmt_vd_m%i_%02i.dot", multiStageIndex, numSplit));
 
-          // Check for read-before-write conflicts in the loop order and counter loop order.
-          // Conflicts in the loop order will assure us that the multi-stage can't be
-          // parallel. A conflict in the counter loop order is more severe and needs the current
-          // multistage be splitted!
-          auto conflict = hasVerticalReadBeforeWriteConflict(&graph, userSpecifiedLoopOrder);
+    if(!splitterIndices.empty()) {
+      auto newMultiStages = multiStage.split(splitterIndices, curLoopOrder);
+      multiStageIt = stencil->getMultiStages().erase(multiStageIt);
```

Comment:
this code is also common. Is it possible to move this out of the lambda, and have the lambda only inside the loop over stages?